File: src/lib/ai/providerChain.ts
import { log } from "@/lib/logger";
import { claudeText, claudeImage } from "./providers/anthropicProvider";
import { geminiText, geminiImage } from "./providers/geminiProvider";
import { opencodeText } from "./providers/opencodeProvider";
import { opencodeImage } from "./providers/ocrProvider";

export const PRIMARY = (process.env.AI_PROVIDER ?? "opencode").toLowerCase();

export type TextFn  = (prompt: string, system: string, maxTokens: number) => Promise<string>;
export type ImageFn = (b64: string, mime: "image/jpeg" | "image/png" | "image/webp", text: string, system: string, maxTokens: number) => Promise<string>;

export function textChain(): TextFn[] {
  const all: TextFn[] = [
    (p, s, t) => claudeText(p, s, t),
    (p, s)    => geminiText(p, s),
    (p, s)    => opencodeText(p, s),
  ];
  if (PRIMARY === "gemini")   return [all[1], all[0], all[2]];
  if (PRIMARY === "opencode") return [all[2], all[0], all[1]];
  return all;
}

export function imageChain(): ImageFn[] {
  const claude:   ImageFn = (b, m, t, s, tok) => claudeImage(b, m, t, s, tok);
  const gemini:   ImageFn = (b, m, t, s)      => geminiImage(b, m, t, s);
  const opencode: ImageFn = (b, m, t, s)      => opencodeImage(b, m, t, s);
  if (PRIMARY === "gemini")   return [gemini,   claude, opencode];
  if (PRIMARY === "opencode") return [opencode, claude, gemini];
  return                              [claude,   gemini, opencode];
}

export async function runChain<T>(chain: Array<() => Promise<T>>, label: string, providers: string[]): Promise<T> {
  let lastErr: unknown;
  for (let i = 0; i < chain.length; i++) {
    const provider = providers[i] ?? `provider-${i}`;
    const t0 = Date.now();
    try {
      const result = await chain[i]();
      log.info("ai", `${label} ok`, { provider, ms: Date.now() - t0 });
      return result;
    } catch (err) {
      lastErr = err;
      log.warn("ai", `${label} failed — trying next`, { provider, ms: Date.now() - t0, err: err instanceof Error ? err.message : String(err) });
    }
  }
  throw lastErr;
}

File: backend/app/ai/provider_chain.py
"""Provider fallback chains — port of src/lib/ai/providerChain.ts."""
import time
from collections.abc import Awaitable, Callable

from app.config import settings
from app.core.logger import log
from app.ai.providers.anthropic_provider import claude_text, claude_image
from app.ai.providers.gemini_provider import gemini_text, gemini_image
from app.ai.providers.opencode_provider import opencode_text
from app.ai.providers.ocr_provider import opencode_image

PRIMARY = (settings.ai_provider or "opencode").lower()

TextFn = Callable[[str, str, int], Awaitable[str]]
ImageFn = Callable[[str, str, str, str, int], Awaitable[str]]


def _order() -> list[str]:
    """Configured provider first, then the others as fallbacks."""
    if PRIMARY == "gemini":
        return ["gemini", "claude", "opencode"]
    if PRIMARY == "opencode":
        return ["opencode", "claude", "gemini"]
    return ["claude", "gemini", "opencode"]


def _configured(name: str) -> bool:
    """Skip a provider we have no usable credential for.

    Attempting one anyway costs a doomed round-trip per call — roughly half a
    second each for claude and gemini — and buries the real failure under
    401s in the log.  opencode authenticates by URL, so it is always eligible.
    """
    if name == "claude":
        return bool(settings.anthropic_api_key)
    if name == "gemini":
        return bool(settings.gemini_api_key)
    return True


def text_chain() -> list[tuple[str, TextFn]]:
    fns: dict[str, TextFn] = {
        "claude": lambda p, s, t: claude_text(p, s, t),
        "gemini": lambda p, s, t: gemini_text(p, s),
        "opencode": lambda p, s, t: opencode_text(p, s),
    }
    return [(n, fns[n]) for n in _order() if _configured(n)]


def image_chain() -> list[tuple[str, ImageFn]]:
    fns: dict[str, ImageFn] = {
        "claude": lambda b, m, t, s, tok: claude_image(b, m, t, s, tok),
        "gemini": lambda b, m, t, s, tok: gemini_image(b, m, t, s),
        "opencode": lambda b, m, t, s, tok: opencode_image(b, m, t, s),
    }
    return [(n, fns[n]) for n in _order() if _configured(n)]


async def run_chain(
    chain: list[Callable[[], Awaitable]], label: str, providers: list[str]
):
    last_err: Exception | None = None
    for i, step in enumerate(chain):
        provider = providers[i] if i < len(providers) else f"provider-{i}"
        t0 = time.time()
        try:
            result = await step()
            log.info("ai", f"{label} ok", {"provider": provider, "ms": int((time.time() - t0) * 1000)})
            return result
        except Exception as err:
            last_err = err
            log.warn(
                "ai",
                f"{label} failed — trying next",
                {"provider": provider, "ms": int((time.time() - t0) * 1000), "err": str(err)},
            )
    raise last_err
